dll/win32/shell32/dialogs/folder_options.cpp

```diff
@@ -701,24 +704,67 @@ Advanced_Compare(const void *x, const void *y)
 {
     ADVANCED_ENTRY *pEntry1 = (ADVANCED_ENTRY *)x;
     ADVANCED_ENTRY *pEntry2 = (ADVANCED_ENTRY *)y;
+
     DWORD dwParentID1 = pEntry1->dwParentID;
     DWORD dwParentID2 = pEntry2->dwParentID;
-    while (dwParentID1 != dwParentID2)
+
+    if (dwParentID1 == dwParentID2)
+        return lstrcmpi(pEntry1->szText, pEntry2->szText);
+
+    DWORD i, m, n;
+    const UINT MAX_DEPTH = 32;
+    ADVANCED_ENTRY *pArray1[MAX_DEPTH];
+    ADVANCED_ENTRY *pArray2[MAX_DEPTH];
+
+    // Make ancestor lists
+    for (i = m = n = 0; i < MAX_DEPTH; ++i)
     {
         ADVANCED_ENTRY *pParent1 = Advanced_GetItem(dwParentID1);
         ADVANCED_ENTRY *pParent2 = Advanced_GetItem(dwParentID2);
         if (!pParent1 && !pParent2)
             break;
-        if (!pParent1 && pParent2)
+
+        if (pParent1)
+        {
+            pArray1[m++] = pParent1;
+            dwParentID1 = pParent1->dwParentID;
+        }
+        if (pParent2)
+        {
+            pArray2[n++] = pParent2;
+            dwParentID2 = pParent2->dwParentID;
+        }
+    }
```

Comment:
If I understand correctly, if you've already hit a null parent on one side you don't need to keep looping. So you could `if (!pParent1 || !pParent2) break;` here instead of the break above.